Quadriga/src/main/java/edu/asu/spring/quadriga/web/LoginController.java

```diff
@@ -70,22 +81,51 @@ public int compare(IProject o1, IProject o2) {
         });
         int size = projectList.size() > 4 ? 4 : projectList.size();
         model.addAttribute("latestProjects", projectList.subList(0, size));
-		return "login";
+        return "login";
 
-	}
+    }
 
-	/**
-	 * Authentication failed. User credentials mismatch causes this request.
-	 * 
-	 * @return		Redirected to login page
-	 */
-	@NoAuthorizationCheck
-	@RequestMapping(value = "/loginfailed", method = RequestMethod.GET)
-	public String loginerror(ModelMap model) {
+    /**
+     * Authentication failed. User credentials mismatch causes this request.
+     * 
+     * @return Redirected to login page
+     */
+    @NoAuthorizationCheck
+    @RequestMapping(value = "/loginfailed", method = RequestMethod.GET)
+    public String loginerror(@RequestParam(value = "socialsignin", required = false) String socialSignInError,
+            ModelMap model) {
+        model.addAttribute("error", "true");
+        return "login";
+    }
 
-		model.addAttribute("error", "true");
-		return "login";
+    @NoAuthorizationCheck
+    @RequestMapping(value = "/sociallogin", method = RequestMethod.GET)
+    public String socialloginerror(@RequestParam(value = "type", required = false) String type, ModelMap model,
+            Locale locale) {
+        String socialSignInErrorMessage = messageSource.getMessage("social.signin.failed", new String[] {}, locale);
+        SocialSignInStatus socialSignInStatus = SocialSignInStatus.getSocialSignInStatus(type);
+        String socialSignInCompletionStatus = "failed";
```

Comment:
if you make this "danger", you don't need an extra if statement in the jsp page (just use for the css class).